dev/AnimatedIcon/AnimatedIcon.cpp

```diff
@@ -0,0 +1,521 @@
+    {
+        if (auto const source = Source())
+        {
+            if (auto const ForegroundSolidColorBrush = Foreground().try_as<winrt::SolidColorBrush>())
+            {
+                source.SetColorProperty(L"Foreground", ForegroundSolidColorBrush.Color());
+            }
+
+            if (auto const visual = source.TryCreateAnimatedVisual(winrt::Window::Current().Compositor()))
+            {
+                m_animatedVisual.set(visual);
+                return visual.RootVisual();
+            }
+            else
+            {
+                m_animatedVisual.set(nullptr);
+                return static_cast<winrt::Visual>(nullptr);
+            }
+        }
+        else
+        {
+            m_animatedVisual.set(nullptr);
+            return static_cast<winrt::Visual>(nullptr);
+        }
+    }();
+
+    if (auto const rootGrid = m_rootGrid.get())
+    {
+        winrt::ElementCompositionPreview::SetElementChildVisual(rootGrid, visual);
+    }
+
+    if (visual)
+    {
+        m_canDisplayPrimaryContent = true;
+        if (auto const rootGrid = m_rootGrid.get())
+        {
+            rootGrid.Children().Clear();
+        }
+        visual.Properties().InsertScalar(L"Progress", 0.0F);
+
+        // Tie the animated visual's Progress property to the player Progress with an ExpressionAnimation.
+        auto compositor = visual.Compositor();
+        auto progressAnimation = compositor.CreateExpressionAnimation(L"_.Progress");
+        progressAnimation.SetReferenceParameter(L"_", m_progressPropertySet);
+        visual.Properties().StartAnimation(L"Progress", progressAnimation);
+    }
+    else
+    {
+        m_canDisplayPrimaryContent = false;
+        if (auto const iconSource = FallbackIconSource())
+        {
+            auto const iconElement = SharedHelpers::MakeIconElementFrom(iconSource);
+            if (auto const rootGrid = m_rootGrid.get())
+            {
+                rootGrid.Children().Clear();
+                rootGrid.Children().InsertAt(0, iconElement);
+            }
+        }
+    }
+}
+
+void AnimatedIcon::OnFallbackIconSourcePropertyChanged(const winrt::DependencyPropertyChangedEventArgs&)
+{
+    if (!m_canDisplayPrimaryContent)
+    {
+        if (auto const iconSource = FallbackIconSource())
+        {
+            auto const iconElement = SharedHelpers::MakeIconElementFrom(iconSource);
+            if (auto const rootGrid = m_rootGrid.get())
+            {
+                rootGrid.Children().Clear();
+                rootGrid.Children().InsertAt(0, iconElement);
+            }
+        }
+    }
+}
+
+void AnimatedIcon::OnForegroundPropertyChanged(const winrt::DependencyObject& sender, const winrt::DependencyProperty& args)
+{
+    if (auto const ForegroundSolidColorBrush = Foreground().try_as<winrt::SolidColorBrush>())
+    {
+        if (auto const source = Source())
+        {
+            source.SetColorProperty(L"Foreground", ForegroundSolidColorBrush.Color());
+        }
+    }
```

Comment:
Duplicated in AnimatedIcon::OnSourcePropertyChanged. Should be a separate method.